[WIP] vSphere Tutorials (#1894)
heptio/velero: site/_posts/2019-09-20-Velero-v1-1-on-vSphere.md

---

> title: Velero v1.1 backing up and restoring apps on vSphere

Reviewer: ```suggestion
title: Backup and Restore Stateless Apps on vSphere using Velero v1.1
```

# image: https://placehold.it/200x200

> excerpt: Velero is an open source tool to safely back up, recover, and migrate Kubernetes clusters and persistent volumes. It works both on premises and in a public cloud. This post demonstrates how it can be used on vSphere.

Reviewer: ```suggestion
excerpt: A How-To guide to run Velero on vSphere.
```

author_name: Cormac Hogan
# author_avatar: https://placehold.it/64x64
categories: ['kubernetes']
# Tag should match author to drive author pages
tags: ['Velero']
---
Velero version 1.1 provides support to backup Kubernetes applications deployed on vSphere. This post will provide detailed information on how to deploy and configure Velero to backup and restore a stateless application (`nginx`) that is running in Kubernetes on vSphere. At this time there is no vSphere plugin for snapshotting stateful applications on vSphere during a Velero backup. In this case, we rely on a third party program called `restic`. However this post does not include an example of how to backup a stateful application. That is available in another tutorial which can be found [here](./2019-20-09-Velero-v1-1-Stateful-Backup-vSphere.md).

## Overview of steps

* Download and extract Velero v1.1

> * Deploy and Configure Minio Object store

Reviewer: ```suggestion
* Deploy and Configure a Minio Object store
```

* Install Velero using the `velero install` command, ensuring that both `restic` support and a Minio `publicUrl` are included
* Run a test backup/restore of a stateless application that has been deployed on upstream Kubernetes

## What this post does not show

* How to install Velero v1.1 with PKS, Pivotal Container Service, on vSphere
* A demonstration on how to do backup/restore of a stateful application (i.e. PVs)

> * The assumption is that the Kubernetes nodes in your cluster have internet access. This guide does not show how to use a local repository for Velero images

Reviewer: ```suggestion
* TThe assumption is that the Kubernetes nodes in your cluster have internet access in order to pull the Velero images. This guide does not show how to add images using a local repository
```

## Download and extract Velero v1.1

> The [Velero v1.1 image can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.

Reviewer: ```suggestion
The [Velero v1.1 binary can be found here](https://github.com/heptio/velero/releases/tag/v1.1.0.). Download and extract it to the desktop where you wish to manage your Velero backups, then copy or move the `velero` binary to somewhere in your $PATH.
```

## Deploy and Configure Minio Object Store as a backup destination

> Velero sends metadata about the objects being backed up to an S3 Object Store. If you do not have an S3 Object Store available, Velero provides the manifest file to create a Minio S3 Object Store on your Kubernetes cluster. This means that all Velero backsup can be kept on-premises.

Reviewer: Besides metadata, we also send the actual contents of the k8s resources. Might be useful to extend the explanation.

Author: Done

```

```bash
$ velero restore create nginx-restore --from-backup nginx-backup
Restore request "nginx-restore" submitted successfully.
Run `velero restore describe nginx-restore` or `velero restore logs nginx-restore` for more details.
```

The following command can be used to examine the restore in detail, and check to see if it has successfully completed.

```bash
$ velero restore describe nginx-restore
Name:         nginx-restore
Namespace:    velero
Labels:       <none>
Annotations:  <none>

Phase:  Completed

Backup:  nginx-backup

Namespaces:
  Included:  *
  Excluded:  <none>

Resources:
  Included:        *
  Excluded:        nodes, events, events.events.k8s.io, backups.velero.io, restores.velero.io, resticrepositories.velero.io
  Cluster-scoped:  auto

Namespace mappings:  <none>

Label selector:  <none>

Restore PVs:  auto
```

## Verify that the restore succeeded

You can see that the restore has now completed. Check to see if the namespace, DaemonSet and service has been restored using the `kubectl` commands shown previously. One item to note is that the `nginx` service may be restored with a new IP address from the LoadBalancer. This is normal.

> Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes I can! Looks like the restore was successful.

Reviewer: ```suggestion
Now let’s see if we can successfully reach our `nginx` web server on that IP address. Yes we can! Looks like the restore was successful.
```

![nginx restored](../img/vsphere-tutorial-icons/nginx-restore-new-ip.png)

> Backups and Restore are now working on Kubernetes deployed on vSphere using Velero v1.1.

Reviewer: ```suggestion
Backups and Restores are now working on Kubernetes deployed on vSphere using Velero v1.1.
```
